feat: Support for quick actions so you can ignore errors more easily

GottemHams opened the pull request:
It supports multiple error identifiers by extending a previous `@phpstan-ignore` comment.

Actually also fixes a crash when encountering a non-ignorable error, due to an undefined variable (https://github.com/Rockstar04/SublimeLinter-contrib-phpstan/issues/21#issuecomment-2641724826 possibly). :>

GottemHams committed:
feat: Support for quick actions so you can ignore errors more easily (actually also fixes a crash when encountering a non-ignorable error)

GottemHams commented on linter.py:
```diff
@@ -19,10 +21,7 @@ def on_activated_async(self):
-    regex = None
     error_stream = lint.STREAM_STDOUT
-    default_type = "error"
-    multiline = False
```
These variables no longer seem to be necessary. At least, I didn't notice any difference with or without them.

kaste replied:
These are indeed unused when you implement parse_output.  Only the regex engine/parser uses them.

GottemHams committed:
chore: Remove some useless code

pierresh commented:
Can you please share an example of PHP code to reproduce the effect (and maybe screenshots of the result)? I am not sure to get the point

GottemHams commented:
I made a video of the hover functionality, screenshots aren't really clear for that one. It's a very silly piece of code of course.

https://github.com/user-attachments/assets/3f3af631-b864-4f17-a9cf-4a8cff80d88e

And if you use Command Palette > `SublimeLinter: Quick Action` when your keyboard cursor is on an offending line, it'll show the same actions:
![command_palette](https://github.com/user-attachments/assets/29261df1-5660-47c9-836f-1de9c330a39b)

Which should behave the same as the hover stuff otherwise.

pierresh commented:
> I made a video of the hover functionality, screenshots aren't really clear for that one. It's a very silly piece of code of course.
> 
>  quick_action.mp4 
> And if you use Command Palette > `SublimeLinter: Quick Action` when your keyboard cursor is on an offending line, it'll show the same actions: ![command_palette](https://private-user-images.githubusercontent.com/20195269/419107939-29261df1-5660-47c9-836f-1de9c330a39b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExNjgyNTMsIm5iZiI6MTc0MTE2Nzk1MywicGF0aCI6Ii8yMDE5NTI2OS80MTkxMDc5MzktMjkyNjFkZjEtNTY2MC00N2M5LTgzNmYtMWRlOWMzMzBhMzliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA1VDA5NDU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxZDE3YTAzZDE3YjYzYjhhOGZlYjhhZWQzZTU2MjM0NWI5NWFkYWMzNzVlZTYyYTJjNjI3MzBkYTAzMTM0YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Hjjkl2XKX4OjcnND3fTDDXXTYbNfHfimAwYX7uyivO0)
> 
> Which should behave the same as the hover stuff otherwise.

Thanks, this looks cool! I will try to review it asap, this week-end at most

kaste commented on linter.py:
```diff
@@ -329,3 +335,41 @@ def find_position_key(self, key, line_content):
+def phpstan_actions_provider(errors, view):
+    def make_action(error):
+        # Newer phpstan versions actually require a specific identifier to ignore, the fallback is just for some backwards compatibility
+        # Note though that without a valid identifier, the on-hover quick actions won't list it (still works via the Command Palette though)
```
> Note though that without a valid identifier, the on-hover quick actions won't list it

That's fixed in master and therefore on the next SublimeLinter release.

kaste commented on linter.py:
```diff
@@ -142,17 +141,24 @@ def find_errors(self, output):
+                if processed_error:
+                    # We'll display some quick actions for ignorable errors later
+                    if 'ignorable' not in error or error['ignorable']:
+                        processed_error['ignore_error'] = error['message']
```
The conditions `'ignorable' not in error or error['ignorable']` looks redundant.

GottemHams replied:
It's mostly for backwards compatibility. If `ignorable` doesn't exist at all then directly accessing `error['ignorable']` will of course cause a `KeyError`.

kaste replied:
But it looks like every error is ignorable because of line 359ff.  (The blank ignore statement) Maybe you don't need `ignore_error`, on line 348 you can also do `"{msg}".format(**error)`.  ❓ 🤷

GottemHams replied:
Not every error is ignorable though, phpstan has a couple of built-in rules that can never be ignored and you can of course implement any non-ignorable custom rule. On older versions the `ignorable` property didn't yet exist (nor did identifiers), which is the reason for the blanket ignore. Newer versions require an identifier even for custom rules as far as I can tell, so it should never trigger the fallback. I can't use `{msg}` because it might also include a tip from phpstan, which can take up quite a bit of extra space and it isn't needed if you want to ignore it anyway.

I actually did realise a little problem though, I've committed another change to the new PR for that.

pierresh committed:
Merge branch 'master' into feat-quick-actions

pierresh commented:
Hello @GottemHams ,

It seems working well, so I merged the PR, but now from time to time I get such feedback when I open a PHP file (but PHPStan then works well).. Do you have the same problem?

<img width="854" alt="image" src="https://github.com/user-attachments/assets/db034d4a-6672-4c5f-a681-5d7eb35e908a" />

kaste commented:
Safer variant is:

```diff
-output = proc
+output = proc.strip()
+if not output:
+    return 
```

GottemHams commented:
@pierresh Apparently that occurs when phpstan throws an error before it even starts scanning, which also means it won't output any JSON at all. For example: you're editing `foo.php` and your `phpstan.neon` actually excludes `foo.php` one way or another, then phpstan will error `[ERROR] No files found to analyse.` when you have that file open in your editor, causing the error panel to show.

https://github.com/Rockstar04/SublimeLinter-contrib-phpstan/pull/25

GottemHams commented:
@pierresh Could you create a new release? :> I never realised this PR was never actually published for Sublime and kept wondering why linting still broke every now and then (the `error_list` problem also fixed by this PR).

I also have one slight improvement for the regex I added originally:

```diff
-        re.compile(r'^Elapsed time: .*?(\r?\nUsed memory: .*)?$', re.MULTILINE),
+        re.compile(r'^(Elapsed time|Used memory): .+', re.MULTILINE),
```

Apparently phpstan doesn't necessarily print both those lines. Stripping the newlines is unnecessary because after all replacements the final string is trimmed anyway.
